File: 986IntervalListIntersections/LC986.py
class Solution:
    def intervalIntersection(self, A: List[List[int]], B: List[List[int]]) -> List[List[int]]:
        if not A or not B:
            return []
        res = []
        i =j =0
        while i < len(A) and j < len(B):
            # if overlap
            st = max(A[i][0], B[j][0])
            en = min(A[i][1], B[j][1])
            if st<= en:
                res.append([st,en])
            
            # remove smallest endpoint interval
            if A[i][1]< B[j][1]:
                i += 1
            else:
                j +=1 
        return res
                

File: 986IntervalListIntersections/LC986.cs
public class Solution {
    public int[][] IntervalIntersection(int[][] A, int[][] B) {
        if (A is null){
            return A;
        }
        if (B is null){
            return B;
        }
        
        List<int[]> res = new List<int[]>();
        int i =0; 
        int j =0;
        while (i < A.Length && j < B.Length){
            int st = Math.Max(A[i][0], B[j][0]);
            int en = Math.Min(A[i][1], B[j][1]);
            if (st <= en){
                res.Add(new int[]{st,en});
            }
            
            if(A[i][1] < B[j][1]){
                i++;
            }
            else{
                j++;
            }
        }
        return res.ToArray();
        
        
    }
}